### PR TITLE
Framework wide color

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/127855
integration test: https://github.com/flutter/engine/pull/54415

This does the preliminary work for implementing wide gamut colors in the Flutter framework.  Here are the following changes:
1) colors now specify a colorspace with which they are to be interpreted
1) colors now store their components as floats to accommodate bit depths more than 8

The storage of this Color class is weird with float/int storage but that is a temporary solution to support a smooth transition.  Here is the plan for landing this:
1) Land this PR
1) Wait for it to roll into the Framework
1) Land https://github.com/flutter/flutter/pull/153938 which will make CupertinoDynamicColor implement Color
1) Land another engine PR that rips out the int storage: https://github.com/flutter/engine/pull/54714

Here are follow up PRs:
1) https://github.com/flutter/engine/pull/54473 - changes DlColor so the wide gamut colors are rendered
1) https://github.com/flutter/engine/pull/54567 - Hooks up these changes to take advantage of wide DlColor
1) https://github.com/flutter/flutter/pull/153319 - the integration test for the framework repo

There are some things that have been left as follow up PRs since they are technically breaking:
1) The math on `lerp` hasn't been updated to take advantage of the higher bit depth
1) `operator==` hasn't been updated to take advantage of the higher bit depth
1) `hashCode` hasn't been updated to take advantage of the higher bit depth
1) `alphaBlend` hasn't been updated to take advantage of the higher bit depth
1) `toString` hasn't been updated to take advantage of the higher bit depth

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/ui/painting/paint.cc
Comment (gaaclarke) [inline lib/ui/painting/paint.cc:79]: Addressed in https://github.com/flutter/engine/pull/54473

**Comment by gaaclarke** [comment]
There is a slight change in the results of alpha blending which is tripping up pixel exact golden tests.  For example in `EmbedderTest.CanRenderSceneWithoutCustomCompositorMetal` the blend of the red and green triangle was (r:85, g:170, b:0), but is now showing up as (r:86, g:169: b:0).

File: lib/ui/painting.dart
Comment (gaaclarke) [inline lib/ui/painting.dart:204]: The design doc had this written as "with()" but dart doesn't allow that name.  ( https://github.com/dart-lang/language/issues/4020 )

**Comment by gaaclarke** [comment]
@johnmccutchan I chose you to review since you have been involved in discussions about this feature and have a wholistic view of engine/framework.  @jonahwilliams I chose you to represent the interests of the engine team.  Let me know if you'd like to hear from someone else.

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:110]: I feel like this deserves a bit more documentation and links from other constructors. For starters: "why would I want to use this constructor over other constructors"?

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:184]: Can be static method

File: lib/ui/painting.dart
Comment (jonahwilliams) [reply lib/ui/painting.dart:234]: I don't love `change`. Maybe `withValues`?

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:437]: You should change the hash code computation in this PR and not as a follow up.

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:426]: You should change the `==` implementation in this PR and not in a follow up. You'll want some sort of fudge factor for the comparisons.

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:1719]: This should document the state of the world on flutter, regardless of hardware support and what happens when it is used on unsupported platforms.

Example: Supported on iOS devices when using the Impeller rendering engine. If used on other platforms, values will be clamped at max sRGB values.

File: lib/ui/painting.dart
Comment (jonahwilliams) [reply lib/ui/painting.dart:110]: Also link to color space enum.

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:1725]: Instead make color space an enhanced enum and give it this field?

Also you can do ColorSpace.values[index] for the reverse.

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:3701]: Can these values be const?

**Review by jonahwilliams — COMMENTED** [review]
What happens if I used a p3 color value after this change lands? Does it get clamped or re-interpreted as sRGB?

**Comment by gaaclarke** [comment]
> What happens if I used a p3 color value after this change lands? Does it get clamped or re-interpreted as sRGB?

It will get clamped.  After https://github.com/flutter/engine/pull/54473 they will be rendered correctly.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart:3706]: Yep!  I wanted to make sure no garbage is generated when performing transforms.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart:1730]: I'm not a fan of this because that would put something that is private into the public API.  We could make it inaccessible but it would still show up to people reading through the enum class.

```dart
enum ColorSpace {
  sRGB(0),
  extendedSRGB(1),
  displayP3(2);

  const ColorSpace(this._index);
  final int _index;
}
```

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart]: Done.  I didn't want to say anything that will go out of date so I mentioned impeller ios as an example and mentioned what happens when it isn't supported.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart]: Done.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart:114]: Done.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart]: `withValues` sounds good to me.  I misremembered thinking that `change` was used elsewhere, hoping to latch onto previous places where we couldn't use `with`.

File: lib/ui/painting.dart
Comment (jonahwilliams) [inline lib/ui/painting.dart:246]: Maybe call this something else...

File: lib/ui/painting.dart
Comment (jonahwilliams) [reply lib/ui/painting.dart:430]: Thinking about this some more, we want ensure the hashCode is equal if the color is equal, so normaliziing to some bit depth (not 256) is probably required here.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart]: What you like, "twisted"?  =). Renamed "updatedComponents".

File: lib/ui/painting.dart
Comment (jonahwilliams) [reply lib/ui/painting.dart:1730]: That seems ... ok to me? but Its 6 of one, half a. dozen of the other, so either approach is fine to me

**Comment by jonahwilliams** [comment]
Clamping seems reasonable. What about gradient colors? The constructors don't take Int32Lists, so we can change the implementation to pass float data now without breaking them. Obvs, atlas and vertices are more complicated.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart:430]: Okay, so I had this all implemented and decided to pull it out and wait for new PR.  The reason why is that I want to separate the PRs for landing support for rendering wide gamut colors and support for wider bit depth for Color components.  The reason is that the wider bit depth in color is the thing that is most likely going to cause a regression in tests.  Rendering wide gamut colors is 3 commits across 2 repos and will be a pain to revert.  I really think these should be landed separately.

My plan is to basically do the same math but use a number like 2^12 to match the bit depth for HDMI 2.0.  On iOS we currently can only render 2^10.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart:441]: Responded to this in https://github.com/flutter/engine/pull/54415#discussion_r1717512260

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart:430]: We could switch the storage back to an int in the meantime if it makes landing this PR easier.

**Comment by gaaclarke** [comment]
> Clamping seems reasonable. What about gradient colors? The constructors don't take Int32Lists, so we can change the implementation to pass float data now without breaking them. Obvs, atlas and vertices are more complicated.

Yea, those are going to be changed after this lands in different PRs.  This is just what is required to make the integration test work.

File: lib/ui/painting.dart
Comment (gaaclarke) [reply lib/ui/painting.dart:430]: In fact maybe it makes sense to switch the storage to always be fixed point.  I could refactor the old code to work from a fixed point where the divisor is 255.  Then we can crank it up to 6000.  The transmission to the lower levels can remain floats.  The benefit is that the change at dart:ui is much simpler.  Having 6000 be the fixed-point divisor would allow us to accurately represent 1/2, 1/4, 1/8, 1/10, 1/16.  It's close to the 2^12 bit rate.

**Comment by gaaclarke** [comment]
The failure in the framework tests can be reproduced with
```
flutter --local-engine=host_debug_unopt_arm64 \
  --local-engine-host=host_debug_unopt_arm64 \
  --local-engine-src-path=/Users/aaclarke/dev/engine/src/ test \
  --name="Cupertino.*focusable\sand" \
  test/material/switch_test.dart 
```

I've traced the failure and it has to do with Paint.color setter and getter which is reading and writing to `ByteData` when going through the trace canvas test matcher.

**Comment by gaaclarke** [comment]
Looks like we have a test too that makes sure the color components are based off of the `.value` result:
```dart
  test('subclass of Color can override value', () {
    const DynamicColorClass color = DynamicColorClass(0xF0E0D0C0);
    expect(color.value, 0xF0E0D0C0);
    // Call base class member, make sure it uses overridden value.
    expect(color.red, 0xE0);
  });
```
https://github.com/gaaclarke/engine/blob/4771d838406fb5dea21a385bf693c384918d9f29/lib/web_ui/test/ui/color_test.dart#L143-L148

**Comment by gaaclarke** [comment]
I was able to eliminate some of the framework test failures by making sure things like `.red` are based on `.value`.  The remaining bugs still seem to be related to Cupertino so possibly another weird subclassing Color behavior.  We should really make that class final.

**Comment by gaaclarke** [comment]
I found the existing error with Cupertino's dynamic color Color subclass.  It was relying on `withAlpha()` which was rewritten to use the float values, but the subclass was using the `.value` override to make that work.  As a result I've deprecated `withOpacity()` and kept it the same definition.

**Comment by gaaclarke** [comment]
I'm dropping this back to a draft until I get all the framework tests passing.

**Comment by gaaclarke** [comment]
I'm having a hell of a time getting the new Color to handle cupertinos subclassing of Color so I'm switching all the maths back to ints as much as possible, while keeping the interface we want.

**Comment by gaaclarke** [comment]
@jonahwilliams The current state of this PR is that the Color class has the API that we want but it is a bit weird since it stores data in an int or a float depending on how it was created.  This was done to support the subclassing of Color in the framework.  My ideal way to land this is to land it like this so that I can get the integration test landed, then going back and migrating the Color subclass away from inheritance, then ripping out the int storage.

The other option is to hold off on this change and forward fix the framework to remove the inheritance, then come back and land this without the double backend storage.

Let me know what you prefer.  It would be nice to have the new api available when editing the framework.

**Comment by jonahwilliams** [comment]
I think we should remove the framework inheritance first and then mark Color as final. Do you have a way to smoke test how disruptive that would be?

**Comment by gaaclarke** [comment]
> I think we should remove the framework inheritance first and then mark Color as final. Do you have a way to smoke test how disruptive that would be?

I'm sifting through it right now.  There is actually 5 subclasses of Color, the cupertino one was the one that gave me the most trouble.

**Comment by jonahwilliams** [comment]
If we want to make color final (giving us room to add/remove members in the future) we can't implement it outside of the engine at all though. How disruptive would it be if CupertinoDynamicColor didn't implement or extend color?

**Comment by jonahwilliams** [comment]
I mean, we can give up on making it final if need be. I would choose whichever option doesn't require a manual engine roll to succeed.

**Comment by gaaclarke** [comment]
> If we want to make color final (giving us room to add/remove members in the future) we can't implement it outside of the engine at all though.

Why's that?  Is there a rule that implementing a sealed class has to be in the same package?

> How disruptive would it be if CupertinoDynamicColor didn't implement or extend color?

Very disruptive.  I mentioned it above but I tried to do this first and it's hard because we are doing runtime reflection.
```dart
void foo(Color x) {
  if (x is CupertinoDynamicColor) {
    //
  }. else {
    //
  }
}
```

The basic steps to remove it would be to always store CupertinoDynamicColor where needed and resolve them when building widgets.  The problem is that they are getting sent through layers that need Color, but then have custom logic based on the reflected type.  It's going to be a difficult knot to untie.

**Comment by jonahwilliams** [comment]
I thought final meant only extended or implemented in same library.

**Comment by gaaclarke** [comment]
> I thought final meant only extended or implemented in same library.

Ahh, you're right, drat.  Well I still think we should move to `implements`.  We can battle with trying to make the class sealed in the future.  Moving to `implements` won't make that any harder or easier and isn't necessary for us to make this work.

**Comment by gaaclarke** [comment]
I authored the PR that will switch the framework to composition.  However, I'm realizing that we won't be able to get the web tests to ever pass on this PR now.  The framework tests will always fail because adding a method to Color will break the framework.  Unless there is a magical way to switch the version of the framework the tests are run against.

**Comment by gaaclarke** [comment]
I talked to jonah offline.  It makes sense to land this, edit the framework, then go back and rip out the int storage.  That way avoids landing things on red or custom engine->framework rolls.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**
